[Gloo][BE] Print stacktrace on collectFullMesh

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98810

Catch error and torch_check it so full C++ stacktrace is printed for
better debug

Differential Revision: [D44860626](https://our.internmc.facebook.com/intern/diff/D44860626/)

rohan-varma committed:
[Gloo][BE] Print stacktrace on collectFullMesh

Catch error and torch_check it so full C++ stacktrace is printed for
better debug

Differential Revision: [D44860626](https://our.internmc.facebook.com/intern/diff/D44860626/)

[ghstack-poisoned]

wanchaol commented on torch/csrc/distributed/c10d/ProcessGroupGloo.cpp:
```diff
@@ -759,7 +759,13 @@ ProcessGroupGloo::ProcessGroupGloo(
+    } catch (const std::runtime_error &e) {
+      auto err = e.what();
+      // TORCH_CHECK to print the cpp stacktrace.
+      TORCH_CHECK(false, c10::str("Gloo connectFullMesh failed with ", err));
```
do we want to call `logAndThrow` instead of manually call `TORCH_CHECK`?

wanchaol reviewed (approved):
lgtm, one suggestion about error handling/rethrowing

rohan-varma committed:
Update on "[Gloo][BE] Print stacktrace on collectFullMesh"


Catch error and torch_check it so full C++ stacktrace is printed for
better debug

Differential Revision: [D44860626](https://our.internmc.facebook.com/intern/diff/D44860626/)

[ghstack-poisoned]

rohan-varma committed:
Update on "[Gloo][BE] Print stacktrace on collectFullMesh"


Catch error and torch_check it so full C++ stacktrace is printed for
better debug

Differential Revision: [D44860626](https://our.internmc.facebook.com/intern/diff/D44860626/)

[ghstack-poisoned]

rohan-varma commented:
@pytorchbot merge -f "CI finished"
